backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CachedLibraries.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.kotlin.backend.konan
+
+import org.jetbrains.kotlin.konan.file.File
+import org.jetbrains.kotlin.konan.library.KonanLibrary
+import org.jetbrains.kotlin.konan.target.CompilerOutputKind
+import org.jetbrains.kotlin.konan.target.KonanTarget
+import org.jetbrains.kotlin.library.uniqueName
+import org.jetbrains.kotlin.utils.addToStdlib.firstNotNullResult
+
+internal class CachedLibraries(
+        private val target: KonanTarget,
+        allLibraries: List<KonanLibrary>,
+        explicitCaches: Map<KonanLibrary, String>,
+        implicitCacheDirectories: List<String>
+) {
+
+    class Cache(val kind: Kind, val path: String) {
+        enum class Kind { DYNAMIC, STATIC }
+    }
+
+    private val allCaches: Map<KonanLibrary, Cache> = allLibraries.mapNotNull { library ->
+        val explicitPath = explicitCaches[library]
+
+        val cache = if (explicitPath != null) {
+            val kind = when {
+                explicitPath.endsWith(target.family.dynamicSuffix) -> Cache.Kind.DYNAMIC
+                explicitPath.endsWith(target.family.staticSuffix) -> Cache.Kind.STATIC
+                else -> error("unexpected cache: $explicitPath")
+            }
+            Cache(kind, explicitPath)
+        } else {
+            implicitCacheDirectories.firstNotNullResult { path ->
+                val dir = File(path).takeIf { it.isDirectory } ?: return@firstNotNullResult null
+                val baseName = "${library.uniqueName}-cache"
```

Comment:
Maybe add target name to mangling?

Reply:
Different targets may have different cache directories instead.
Also I'm not sure that target is the only compiler option to take into account here.